src/main/java/net/dries007/tfc/world/feature/vein/PipeVeinFeature.java

```diff
@@ -8,36 +8,63 @@
 
 import java.util.Random;
 
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.MutableBoundingBox;
 
 import com.mojang.serialization.Codec;
 
-public class PipeVeinFeature extends VeinFeature<PipeVeinConfig, Vein>
+public class PipeVeinFeature extends VeinFeature<PipeVeinConfig, PipeVeinFeature.PipeVein>
 {
     public PipeVeinFeature(Codec<PipeVeinConfig> codec)
     {
         super(codec);
     }
 
     @Override
-    protected MutableBoundingBox getBoundingBox(PipeVeinConfig config)
+    protected MutableBoundingBox getBoundingBox(PipeVeinConfig config, PipeVein vein)
     {
-        return new MutableBoundingBox(-config.getRadius(), -config.getSize(), -config.getRadius(), config.getRadius(), config.getSize(), config.getRadius());
+        int radius = config.getRadius();
+        int skew = vein.skew;
+        return new MutableBoundingBox(-radius - skew, -config.getSize(), -radius - skew, radius + skew, config.getSize(), radius + skew);
     }
 
     @Override
-    protected float getChanceToGenerate(int x, int y, int z, Vein vein, PipeVeinConfig config)
+    protected float getChanceToGenerate(int x, int y, int z, PipeVein vein, PipeVeinConfig config)
     {
-        if (Math.abs(y) < config.getSize() && (x * x) + (z * z) < config.getRadius() * config.getRadius())
+        final double yScaled = (double) y / config.getSize();
+        x += vein.skew * MathHelper.cos(vein.angle) * yScaled;
+        z += vein.skew * MathHelper.sin(vein.angle) * yScaled;
```

Comment:
These can be precomputed on the vein itself - i.e. from `vein.angle` and `vein.skew`, compute `vein.skewX` and `vein.skewZ`. Since this will be called a lot more often.